File: backend/definition/validation/validate_skill.py
from __future__ import annotations

import re
from typing import Any, Literal

from runtime.guards import is_valid_guard_syntax
from definition.behavior.compile import compile_behavior
from definition.behavior.validate import validate_behavior_graph

AgentIssueSeverity = Literal["error", "warning", "info"]
_SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+$")


def _issue(
    issue_id: str,
    severity: AgentIssueSeverity,
    code: str,
    message: str,
    href: str | None = None,
) -> dict[str, Any]:
    return {
        "id": issue_id,
        "severity": severity,
        "code": code,
        "message": message,
        "href": href,
    }


def _reachable_states(skill: dict[str, Any]) -> set[str]:
    states = skill.get("states", [])
    state_ids = {state["id"] for state in states if state.get("id")}
    initial = skill.get("initial")
    if not initial or initial not in state_ids:
        return set()

    reachable: set[str] = {initial}
    queue = [initial]
    while queue:
        current = queue.pop(0)
        state = next((item for item in states if item.get("id") == current), None)
        if state is None:
            continue
        for transition in state.get("transitions", []):
            target = transition.get("to")
            if target in state_ids and target not in reachable:
                reachable.add(target)
                queue.append(target)
    return reachable


def validate_skill(
    skill: dict[str, Any],
    *,
    action_ids: set[str],
    href: str | None = None,
) -> list[dict[str, Any]]:
    issues: list[dict[str, Any]] = []
    skill_id = skill.get("id", "")
    skill_name = skill.get("name", skill_id)
    version = str(skill.get("version") or "0.1.0")

    if not _SEMVER_RE.match(version):
        issues.append(
            _issue(
                f"skill.{skill_id}.bad-version",
                "warning",
                "invalid_skill_version",
                f"Skill «{skill_name}»: version «{version}» не SemVer (ожидается X.Y.Z)",
                href,
            )
        )

    behavior = skill.get("behavior") if isinstance(skill.get("behavior"), dict) else None
    has_behavior = bool(behavior and behavior.get("nodes"))
    if has_behavior:
        for issue in validate_behavior_graph(behavior, action_ids=action_ids, skill_id=skill_id):
            issues.append(
                _issue(
                    f"skill.{skill_id}.behavior.{issue.get('code')}.{issue.get('nodeId') or 'graph'}",
                    issue.get("severity") or "error",
                    str(issue.get("code") or "behavior"),
                    f"Skill «{skill_name}»: {issue.get('message')}",
                    href,
                )
            )
        compiled = compile_behavior(behavior, action_ids=action_ids)
        if compiled["ok"] and compiled["artifact"]:
            skill = {
                **skill,
                "states": compiled["artifact"]["states"],
                "initial": compiled["artifact"]["initial"],
            }

    state_ids = {state["id"] for state in skill.get("states", []) if state.get("id")}

    if not skill.get("states"):
        issues.append(
            _issue(
                f"skill.{skill_id}.empty",
                "warning",
                "skill_empty",
                f"Skill «{skill_name}» без states",
                href,
            )
        )
        return issues

    initial = skill.get("initial")
    if not initial:
        issues.append(
            _issue(
                f"skill.{skill_id}.no-initial",
                "error",
                "missing_initial",
                f"Skill «{skill_name}»: не задан initial state",
                href,
            )
        )
    elif initial not in state_ids:
        issues.append(
            _issue(
                f"skill.{skill_id}.bad-initial",
                "error",
                "invalid_initial",
                f"Skill «{skill_name}»: initial «{initial}» отсутствует среди states",
                href,
            )
        )

    reachable = _reachable_states(skill)
    for state in skill.get("states", []):
        state_id = state.get("id", "")
        if state_id and state_id not in reachable and state_id != initial:
            issues.append(
                _issue(
                    f"skill.{skill_id}.state.{state_id}.unreachable",
                    "warning",
                    "unreachable_state",
                    f"Skill «{skill_name}» / {state_id}: state недостижим из initial",
                    href,
                )
            )

        if not state.get("final"):
            has_outgoing = bool(state.get("transitions")) or bool(state.get("onEnter"))
            if not has_outgoing and state_id in reachable:
                issues.append(
                    _issue(
                        f"skill.{skill_id}.state.{state_id}.dead-end",
                        "warning",
                        "dead_end_state",
                        f"Skill «{skill_name}» / {state_id}: не-final state без переходов",
                        href,
                    )
                )

        for action_id in state.get("onEnter", []):
            if action_id not in action_ids:
                issues.append(
                    _issue(
                        f"skill.{skill_id}.state.{state_id}.on_enter.{action_id}",
                        "error",
                        "unknown_action",
                        f"Skill «{skill_name}» / {state_id}: on_enter ссылается на неизвестный Action «{action_id}»",
                        href,
                    )
                )

        events_seen: dict[str, list[str]] = {}
        for transition in state.get("transitions", []):
            transition_id = transition.get("id", "")
            event = (transition.get("event") or "").strip()
            guard = transition.get("guard", "")

            if not event:
                issues.append(
                    _issue(
                        f"skill.{skill_id}.transition.{transition_id}.event",
                        "warning",
                        "empty_event",
                        f"Skill «{skill_name}» / {state_id}→{transition.get('to') or '?'}: пустой event",
                        href,
                    )
                )

            if guard and not is_valid_guard_syntax(guard):
                issues.append(
                    _issue(
                        f"skill.{skill_id}.transition.{transition_id}.guard",
                        "error",
                        "invalid_guard_syntax",
                        f"Skill «{skill_name}» / {state_id}: guard «{guard}» синтаксически невалиден",
                        href,
                    )
                )

            to_state = transition.get("to")
            if not to_state or to_state not in state_ids:
                issues.append(
                    _issue(
                        f"skill.{skill_id}.transition.{transition_id}.to",
                        "error",
                        "invalid_transition_target",
                        f"Skill «{skill_name}» / {state_id}: переход ведёт в неизвестный state «{to_state or '—'}»",
                        href,
                    )
                )

            for action_id in transition.get("actions", []):
                if action_id not in action_ids:
                    issues.append(
                        _issue(
                            f"skill.{skill_id}.transition.{transition_id}.action.{action_id}",
                            "error",
                            "unknown_action",
                            f"Skill «{skill_name}» / {state_id}: transition ссылается на неизвестный Action «{action_id}»",
                            href,
                        )
                    )

            if event:
                key = f"{state_id}:{event}"
                events_seen.setdefault(key, []).append(guard.strip() or "")

        for key, guards in events_seen.items():
            empty_guards = [item for item in guards if not item]
            if len(empty_guards) > 1:
                _, event_name = key.split(":", 1)
                issues.append(
                    _issue(
                        f"skill.{skill_id}.state.{state_id}.event.{event_name}.ambiguous",
                        "warning",
                        "ambiguous_guards",
                        f"Skill «{skill_name}» / {state_id}: event «{event_name}» имеет несколько handlers без guard",
                        href,
                    )
                )

    for param in skill.get("params", []):
        name = str(param.get("name") or "").strip()
        if not name:
            issues.append(
                _issue(
                    f"skill.{skill_id}.param.empty",
                    "warning",
                    "empty_param_name",
                    f"Skill «{skill_name}»: param без name",
                    href,
                )
            )

    return issues

File: frontend/src/features/skills/utils/validateSkill.ts
import type {
  AgentIssue,
  AgentIssueSeverity,
  SkillIssueLocator,
} from "@/features/agents/utils/validateAgent";
import type { Skill } from "@/features/skills/types/skill";
import { compileBehavior } from "@/features/skills/utils/behaviorCompile";
import { skillHasBehavior } from "@/features/skills/utils/behaviorGraph";
import { validateBehaviorGraph } from "@/features/skills/utils/behaviorValidate";

const SEMVER_PATTERN = /^\d+\.\d+\.\d+$/;
const GUARD_PATTERN = /^[\w.]+\s*(==|!=|>=|<=|>|<)\s*.+$/;

function issue(
  id: string,
  severity: AgentIssueSeverity,
  code: string,
  message: string,
  options?: { href?: string; locator?: SkillIssueLocator; actionId?: string },
): AgentIssue {
  return {
    id,
    severity,
    code,
    message,
    href: options?.href,
    locator: options?.locator,
    actionId: options?.actionId,
  };
}

function reachableStates(skill: Skill): Set<string> {
  const stateIds = new Set(skill.states.map((state) => state.id));
  const initial = skill.initial;
  if (!initial || !stateIds.has(initial)) return new Set();

  const reachable = new Set<string>([initial]);
  const queue = [initial];
  while (queue.length > 0) {
    const current = queue.shift()!;
    const state = skill.states.find((item) => item.id === current);
    if (!state) continue;
    for (const transition of state.transitions) {
      if (stateIds.has(transition.to) && !reachable.has(transition.to)) {
        reachable.add(transition.to);
        queue.push(transition.to);
      }
    }
  }
  return reachable;
}

export function isValidGuardSyntax(guard: string): boolean {
  const expression = guard.trim();
  if (!expression) return true;
  return GUARD_PATTERN.test(expression);
}

export function validateSkill(skill: Skill, actionIds: Set<string>, href?: string): AgentIssue[] {
  const issues: AgentIssue[] = [];
  const skillLocator: SkillIssueLocator = { kind: "skill" };

  if (!SEMVER_PATTERN.test(skill.version)) {
    issues.push(
      issue(
        `skill.${skill.id}.bad-version`,
        "warning",
        "invalid_skill_version",
        `Skill «${skill.name}»: version «${skill.version}» не SemVer (ожидается X.Y.Z)`,
        { href, locator: skillLocator },
      ),
    );
  }

  let working = skill;
  if (skillHasBehavior(skill.behavior)) {
    for (const item of validateBehaviorGraph(skill.behavior, { actionIds })) {
      issues.push(
        issue(
          `skill.${skill.id}.behavior.${item.code}.${item.nodeId ?? "graph"}`,
          item.severity,
          item.code,
          `Skill «${skill.name}»: ${item.message}`,
          { href, locator: item.nodeId ? { kind: "node", nodeId: item.nodeId } : skillLocator },
        ),
      );
    }
    const compiled = compileBehavior(skill.behavior, { actionIds });
    if (compiled.ok) {
      working = {
        ...skill,
        states: compiled.artifact.states,
        initial: compiled.artifact.initial,
      };
    }
  }

  const stateIds = new Set(working.states.map((state) => state.id));
  if (working.states.length === 0) {
    issues.push(
      issue(
        `skill.${skill.id}.empty`,
        "warning",
        "skill_empty",
        `Skill «${skill.name}» без states`,
        { href, locator: skillLocator },
      ),
    );
    return issues;
  }

  if (!working.initial) {
    issues.push(
      issue(
        `skill.${skill.id}.no-initial`,
        "error",
        "missing_initial",
        `Skill «${skill.name}»: не задан initial state`,
        { href, locator: skillLocator },
      ),
    );
  } else if (!stateIds.has(working.initial)) {
    issues.push(
      issue(
        `skill.${skill.id}.bad-initial`,
        "error",
        "invalid_initial",
        `Skill «${skill.name}»: initial «${working.initial}» отсутствует среди states`,
        { href, locator: skillLocator },
      ),
    );
  }

  const reachable = reachableStates(working);
  for (const state of working.states) {
    const stateLocator: SkillIssueLocator = { kind: "state", stateId: state.id };

    if (!reachable.has(state.id) && state.id !== working.initial) {
      issues.push(
        issue(
          `skill.${skill.id}.state.${state.id}.unreachable`,
          "warning",
          "unreachable_state",
          `Skill «${skill.name}» / ${state.id}: state недостижим из initial`,
          { href, locator: stateLocator },
        ),
      );
    }

    if (
      !state.final &&
      state.transitions.length === 0 &&
      state.onEnter.length === 0 &&
      reachable.has(state.id)
    ) {
      issues.push(
        issue(
          `skill.${skill.id}.state.${state.id}.dead-end`,
          "warning",
          "dead_end_state",
          `Skill «${skill.name}» / ${state.id}: не-final state без переходов`,
          { href, locator: stateLocator },
        ),
      );
    }

    for (const actionId of state.onEnter) {
      if (!actionIds.has(actionId)) {
        issues.push(
          issue(
            `skill.${skill.id}.state.${state.id}.on_enter.${actionId}`,
            "error",
            "unknown_action",
            `Skill «${skill.name}» / ${state.id}: on_enter ссылается на неизвестный Action «${actionId}»`,
            { href, locator: stateLocator, actionId },
          ),
        );
      }
    }

    const eventsSeen = new Map<string, string[]>();
    for (const transition of state.transitions) {
      const transitionLocator: SkillIssueLocator = {
        kind: "transition",
        stateId: state.id,
        transitionId: transition.id,
      };

      if (!transition.event.trim()) {
        issues.push(
          issue(
            `skill.${skill.id}.transition.${transition.id}.event`,
            "warning",
            "empty_event",
            `Skill «${skill.name}» / ${state.id}→${transition.to || "?"}: пустой event`,
            { href, locator: transitionLocator },
          ),
        );
      }

      if (transition.guard && !isValidGuardSyntax(transition.guard)) {
        issues.push(
          issue(
            `skill.${skill.id}.transition.${transition.id}.guard`,
            "error",
            "invalid_guard_syntax",
            `Skill «${skill.name}» / ${state.id}: guard «${transition.guard}» синтаксически невалиден`,
            { href, locator: transitionLocator },
          ),
        );
      }

      if (!transition.to || !stateIds.has(transition.to)) {
        issues.push(
          issue(
            `skill.${skill.id}.transition.${transition.id}.to`,
            "error",
            "invalid_transition_target",
            `Skill «${skill.name}» / ${state.id}: переход ведёт в неизвестный state «${transition.to || "—"}»`,
            { href, locator: transitionLocator },
          ),
        );
      }

      for (const actionId of transition.actions) {
        if (!actionIds.has(actionId)) {
          issues.push(
            issue(
              `skill.${skill.id}.transition.${transition.id}.action.${actionId}`,
              "error",
              "unknown_action",
              `Skill «${skill.name}» / ${state.id}: transition ссылается на неизвестный Action «${actionId}»`,
              { href, locator: transitionLocator, actionId },
            ),
          );
        }
      }

      const event = transition.event.trim();
      if (event) {
        const key = `${state.id}:${event}`;
        const guards = eventsSeen.get(key) ?? [];
        guards.push(transition.guard.trim());
        eventsSeen.set(key, guards);
      }
    }

    for (const [key, guards] of eventsSeen.entries()) {
      const emptyGuards = guards.filter((value) => !value);
      if (emptyGuards.length > 1) {
        const eventName = key.split(":")[1] ?? key;
        issues.push(
          issue(
            `skill.${skill.id}.state.${state.id}.event.${eventName}.ambiguous`,
            "warning",
            "ambiguous_guards",
            `Skill «${skill.name}» / ${state.id}: event «${eventName}» имеет несколько handlers без guard`,
            { href, locator: stateLocator },
          ),
        );
      }
    }
  }

  skill.params.forEach((param, index) => {
    if (!param.name.trim()) {
      issues.push(
        issue(
          `skill.${skill.id}.param.${index}.empty`,
          "warning",
          "empty_param_name",
          `Skill «${skill.name}»: param без name`,
          { href, locator: { kind: "param", index } },
        ),
      );
    }
  });

  return issues;
}
